### ISSUE TITLE
example using microphone

### ISSUE THREAD

**Issue by AwokeKnowing** [huggingface/distil-whisper#4]
let's face it. these models are developed with static datasets but a primary use case is streaming audio transcription.

Please include a microphone-based demo (or suffer 1000 github issues begging for it, see other whisper repos)

**Comment by AwokeKnowing** [comment]
@sanchit-gandhi well the reality is people need it to run 'online' and latency in getting the text to an LLM is important, but I have not seen a standard effective solution for that (detecting end of utterance). So I strongly suggest you at least do a simple '1 second of silence' implementation which will appease the masses.  Or possibly a 'hold the spacebar down while speaking' implementation.

**Comment by AwokeKnowing** [comment]
well I don't think this repo needs to solve the mic to utterance latency problem.  I hope a few general solutions emerge which can be plugged in to the STT models.  (ai predicted end of utterance, as part of hf pipeline?) But for now, I think there should be an example where you are talking and the live transcription in showing, as this is the best way to evaluate the model in practice.   It would be nice to have both a python and js version.  

When people see a model like this, they want to clone it, run a script and start talking.  If that is there, it's likely to be a go-to repo.

**Comment by AwokeKnowing** [comment]
yeah that would be perfect. A huge boost for this repo if people can clone, run a setup, and start talking and seeing how it works we just webcam audio.

**Comment by kristiankielhofner** [comment]
@AwokeKnowing - We also have a "nearly live" implementation that [uses WebRTC from a browser](https://github.com/toverainc/willow-ts-client) to stream to our [Willow Inference Server](https://github.com/toverainc/willow-inference-server) which uses ctranslate2 and a variety of other performance optimizations we have implemented. It's unique in a variety of ways:

- Once the connection is open and ICE, etc has been negotiated we pause the audio track until you start recording. This allows for long-running sessions over days or longer that don't consume any bandwidth other than necessary bare-minimum exchanges to keep the session active in the browser, any NAT devices, and the inference server. We have users that have left their sessions open for weeks.

- When you start recording we resume the track and begin streaming audio. When you stop recording the WebRTC audio track is paused again and the received audio server side is passed from buffer to whisper.

- The results are sent back via the WebRTC datachannel and displayed in the browser.

You can watch the (old) [demo video](https://www.youtube.com/watch?v=PxCO5eONqSQ) - which is actually slow compared to what we currently have implemented and it will only get faster with distill-whisper.

Distil-whisper coming as soon as we get our hands on it!

**Comment by sanchit-gandhi** [comment]
Okay here's a quick PoC - works best if you have a GPU locally for faster inference. Also make sure you have the package [FFmpeg](https://ffmpeg.org) installed so that you can listen to your device mic. Over the next few days, I'll make a [Gradio](https://www.gradio.app) demo out of this to showcase it end-to-end.

```python
import torch
from transformers import AutoModelForSpeechSeq2Seq, AutoProcessor, pipeline
from transformers.pipelines.audio_utils import ffmpeg_microphone_live
import sys

device = "cuda:0" if torch.cuda.is_available() else "cpu"
torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32

model_id = "distil-whisper/distil-medium.en"

model = AutoModelForSpeechSeq2Seq.from_pretrained(
    model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=True, use_safetensors=True
)
model.to(device)

processor = AutoProcessor.from_pretrained(model_id)

transcriber = pipeline(
    "automatic-speech-recognition",
    model=model,
    tokenizer=processor.tokenizer,
    feature_extractor=processor.feature_extractor,
    max_new_tokens=128,
    torch_dtype=torch_dtype,
    device=device,
)

def transcribe(chunk_length_s=20.0, stream_chunk_s=1.0):
    sampling_rate = transcriber.feature_extractor.sampling_rate

    mic = ffmpeg_microphone_live(
        sampling_rate=sampling_rate,
        chunk_length_s=chunk_length_s,
        stream_chunk_s=stream_chunk_s,
    )

    print("Start speaking...")
    for item in transcriber(mic, generate_kwargs={"max_new_tokens": 128}):
        sys.stdout.write("\033[K")
        print(item["text"], end="\r")
        if not item["partial"][0]:
            break

    return item["text"]

transcribe()
```

You can adjust the chunk length based on how real-time you need the transcription to be. Using a smaller `stream_chunk_s` lends itself to more real-time speech recognition, since we divide our input audio into smaller chunks and transcribe them on the fly. However, this comes at the expense of poorer accuracy, since there’s less context for the model to infer from.

Note also that the function `ffmpeg_microphone_live` has been reported to have some bugs on Windows/Mac. If it's not registering the microphone, there are some monkey patch solutions here: https://github.com/huggingface/transformers/issues/25183#issuecomment-1769312607

If you're running on CPU only, we might need a different implementation for this to work fast enough (e.g. in Rust).

**Comment by AnkushMalaker** [comment]
@sanchit-gandhi How do you suggest a rust port is going to solve the inference speed issue on CPU only systems?
I assumed the bottleneck is data processing through the model, which a rust port won't fix. Or perhaps you're suggesting utilizing multiple cores for some sort of parallelization? At that point, ctranslate2 should be sufficient?

**Comment by sanchit-gandhi** [comment]
`distil-medium.en` works extremely well in particular with Whisper cpp: https://github.com/huggingface/distil-whisper#exporting-to-other-libraries

In a toy benchmark, it's about 4x faster than `large-v2`. This is the best performing CPU-only export I've come across for Distil-Whisper so far!

**Comment by ajeema** [comment]
if it helps, here is a quick FastAPI implementation I just did

`from fastapi import FastAPI, File, UploadFile, HTTPException
from pydantic import BaseModel
import torch
from transformers import AutoModelForSpeechSeq2Seq, AutoProcessor, pipeline
from io import BytesIO
import soundfile as sf
import librosa

app = FastAPI()

class TranscriptionResponse(BaseModel):
    text: str

# Setup CUDA device and data type
device = "cuda:0" if torch.cuda.is_available() else "cpu"
torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32

# Load the model and processor
model_id = "distil-whisper/distil-small.en"
model = AutoModelForSpeechSeq2Seq.from_pretrained(
    model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=True, use_safetensors=True
).to(device)
processor = AutoProcessor.from_pretrained(model_id)

# Define ASR pipelines for short and long-form processing
short_form_pipe = pipeline(
    "automatic-speech-recognition",
    model=model,
    tokenizer=processor.tokenizer,
    feature_extractor=processor.feature_extractor,
    max_new_tokens=128,
    torch_dtype=torch_dtype,
    device=device
)

long_form_pipe = pipeline(
    "automatic-speech-recognition",
    model=model,
    tokenizer=processor.tokenizer,
    feature_extractor=processor.feature_extractor,
    max_new_tokens=128,
    chunk_length_s=25,
    batch_size=16,
    torch_dtype=torch_dtype,
    device=device
)

@app.post("/transcribe", response_model=TranscriptionResponse)
async def transcribe_audio(file: UploadFile = File(...)):
    try:
        # Read and preprocess audio
        audio_bytes = await file.read()
        buffer = BytesIO(audio_bytes)
        buffer.seek(0)
        audio_input, sr_original = sf.read(buffer, dtype='float32')

        if sr_original != 16000:
            audio_input = librosa.resample(audio_input, orig_sr=sr_original, target_sr=16000)

        duration = len(audio_input) / 16000  # Duration in seconds

        if duration <= 30:
            print("Using short-form processing")
            results = short_form_pipe(audio_input)
        else:
            print("Using long-form processing")
            results = long_form_pipe(audio_input)

        transcription = results['text'] if 'text' in results else "No transcription found"

        return TranscriptionResponse(text=transcription)
    except Exception as e:
        print(f"An error occurred: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))

if __name__ == "__main__":
    import uvicorn
    uvicorn.run("app:app", host="0.0.0.0", port=8000, workers=4)

`

and tested it with this

`import requests

def upload_file(file_path):
    url = 'http://localhost:8000/transcribe'
    with open(file_path, 'rb') as f:
        files = {'file': (file_path, f)}
        response = requests.post(url, files=files)
        return response.text

file_path = /path/to/file.mp3'
response = upload_file(file_path)
print(response)
`
